homeassistant/components/camera/__init__.py

```diff
@@ -0,0 +1,1133 @@
+        response = requests.get(
+            self.still_image_url,
+            auth=(self.username, self.password), stream=stream)
+
+        return response
+
+    @property
+    def name(self):
+        """ Return the name of this device """
+        if self.device_info.get('name'):
+            return self.device_info.get('name')
+        else:
+            return super().name
+
+    @property
+    def state(self):
+        """ Returns the state of the entity. """
+        seconds_since_last_motion = (
+            datetime.datetime.now() -
+            self._last_motion_detected).total_seconds()
+
+        if (self._is_detecting_motion and
+                seconds_since_last_motion > EVENT_GAP_THRESHOLD):
+            self._is_detecting_motion = False
+
+        if self._is_detecting_motion:
+            return STATE_MOTION_DETECTED
+        elif self.is_recording:
+            return STATE_RECORDING
+        elif self.is_streaming:
+            return STATE_STREAMING
+        elif self._is_motion_detection_enabled:
+            return STATE_ARMED
+        else:
+            return STATE_IDLE
+
+    @property
+    def state_attributes(self):
+        """ Returns optional state attributes. """
+        attr = super().state_attributes
+        attr['model_name'] = self.device_info.get('model', 'generic')
+        attr['brand'] = self.device_info.get('brand', 'generic')
+        attr['still_image_url'] = '/api/camera_proxy/' + self.entity_id
+        attr[ATTR_ENTITY_PICTURE] = (
+            '/api/camera_proxy/' +
+            self.entity_id + '?time=' +
+            str(time.time()))
+        attr['stream_url'] = '/api/camera_proxy_stream/' + self.entity_id
+        motion_time = self._last_motion_detected.strftime('%Y-%m-%d %H-%M-%S')
+        attr['last_motion_time'] = motion_time
+        attr['last_connected_address'] = self.last_connected_address
+        attr['child_entities'] = self._action_entities
+        attr['is_recording'] = self._is_recording
+        attr['is_taking_snapshot'] = self.is_taking_snapshot
+
+        attr.update(self.function_attributes)
+
+        return attr
+
+    @property
+    def still_image_url(self):
+        """ This should be implemented by different camera models. """
+        return self.base_url + self._still_image_url
+
+    def enable_motion_detection(self):
+        """ Enable the motion detection settings for the camera.
+            This should be overridden in derived classes """
+        if not self.is_motion_detection_supported:
+            return False
+        if self.is_motion_detection_enabled:
+            return True
+
+    def disable_motion_detection(self):
+        """ Disable the motion detection settings for the camera.
+            This should be overridden in derived classes """
+        if not self.is_motion_detection_supported:
+            return False
+        if self.is_motion_detection_enabled:
+            return True
+
+    def set_ftp_details(self):
```

Comment:
I would like to propose to break out the ftp functionality. This will make certain logic reusable and slim down the camera component significant, which would make it easier for new contributors to get up to speed.

First component to create would be a directory watcher component that fires events when it detects changes to a local folder (for example based on [watchdog](https://github.com/gorakhargosh/watchdog/)).

We could then decide to have the ftp server be a separate component or users could choose to use any ftp server that they might have already running on their system.